src/arrays/arraymath.jl

```diff
@@ -135,35 +130,33 @@ end
 # exponential of dense matrix
 function Base.expm(qarr::AbstractQuMatrix)
     fc = expm(full(rawcoeffs(qarr)))
-    QAT = similar_type(typeof(qarr))
+    QAT = similar_type(qarr)
     return QAT(fc, bases(qarr))
 end
 #Base.expm(ct::CTranspose) = expm(ct.qarr)'
 
-
 ##################
 # Tensor Product #
 ##################
 
 # General tensor product definitions for orthonormal bases
-function tensor{B<:OrthonormalBasis,T1,T2,N}(qarr1::AbstractQuArray{B,T1,N}, qarr2::AbstractQuArray{B,T2,N})
+function tensor{B,T1,T2,N}(qarr1::AbstractQuArray{B,T1,N}, qarr2::AbstractQuArray{B,T2,N})
     tc = kron(coeffs(qarr1), coeffs(qarr2))
-    QAT = similar_type(promote_type(typeof(qarr1), typeof(qarr2))) 
+    QAT = similar_type(qarr1, qarr2)
     return QAT(tc, map(tensor, bases(qarr1), bases(qarr2)))
 end
 
-function tensor{B<:OrthonormalBasis,T1,T2,N}(qarr1::CTranspose{B,T1,N}, qarr2::CTranspose{B,T2,N})
-    return QuArray(kron(rawcoeffs(qarr1), rawcoeffs(qarr2)),
-                   map(tensor, rawbases(qarr1), rawbases(qarr2)))'
-end
+# defined to resolve ambiguity warnings
+tensor{B}(ct1::DualVector{B}, ct2::DualVector{B}) = tensor(ct1.qarr, ct2.qarr)'
+tensor{B}(ct1::CTranspose{B}, ct2::CTranspose{B}) = tensor(ct1.qarr, ct2.qarr)'
 
-function tensor{B<:OrthonormalBasis}(ket::QuVector{B}, bra::DualVector{B})
+function tensor{B}(ket::AbstractQuVector{B}, bra::DualVector{B})
     return QuArray(kron(coeffs(ket), coeffs(bra)),
```

Comment:
Good catch. I fixed it, and in the process of doing so uncovered an issue. 

I found out that n-ary calls of `tensor` weren't dispatching properly because of the restriction that all arguments share a basis type. Since we parameterize the number of factors in our basis types, computing things like the following wouldn't work:

```
a = QuArray(rand(Complex{Float64},4), FiniteBasis(2,2))
b = QuArray(rand(Complex{Float64},1), FiniteBasis(1,1))
c = QuArray(rand(Complex{Float64},3))

ab = tensor(a,b)
tensor(ab, c) # fails, ab has a different basis type than c since the number of factors are different
```

After removing the basis type restriction, I added a test similar to the above. It's currently failing  because type promotion for bases of different factor length is ambiguous:

```
julia> promote_type(typeof(ab), typeof(c))
ERROR: type: QuArray: in B, expected B<:AbstractBasis{S<:AbstractStructure}, got Type{FiniteBasis{S,N}}
 in promote_type at promotion.jl:110
```

I don't know what the right decision is here...perhaps we should consider abandon parameterizing factor length for basis types? Will any functions in the future require factor length as a parameter for type stability? Otherwise, we can just pull it from instances using the `nfactors` function (requiring it to be defined for all `B<:AbstractBasis`), and treat factor length as a value.